File: butcher/src/confy/git.zig
//:_______________________________________________________________________
//  ᛝ confy  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU LGPLv3 or later  :
//:_______________________________________________________________________
//! @fileoverview Confy's Git Management tools
//______________________________________________|
const Git = @This();
// @deps std
const std = @import("std");
// @deps zstd
const zstd = @import("../lib/zstd.zig");
const cstr = zstd.cstr;


//______________________________________
/// @descr Describes metadata/information about a Git repository, and provides tool to manage it.
pub const Info = struct {
  baseURL :cstr= "https://github.com",
  owner   :cstr,
  repo    :cstr,

  //______________________________________
  const Templ = "{s}/{s}/{s}";
  /// @descr Provides the functionality required for {@link Git.Info} objects to work with default zig formatting functions.
  pub fn format (G :*const Git.Info, comptime _:[]const u8, _:std.fmt.FormatOptions, writer :anytype) !void {
    try writer.print(Git.Info.Templ, .{G.baseURL, G.owner, G.repo});
  }

  //______________________________________
  /// @descr Returns a formatted URL of the given {@arg G} {@link Git.Info} object.
  /// @note
  ///  Allocates a {@link cstr} owned by the caller.
  ///  Use the provided {@link Git.Info.format} function if you need/prefer to not allocate.
  pub fn URL (G :*const Git.Info, A :std.mem.Allocator) cstr {
    return std.fmt.allocPrint(A, "{s}", .{G});
  }
};

File: src/confy/tools/git.nim
#:______________________________________________________________________
#  ᛝ confy  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU GPLv3 or later  :
#:______________________________________________________________________
##! @fileoverview Confy's Git Management tools
#______________________________________________|
from std/strformat import fmt

#_______________________________________
# @section Comptime Defaults
#_____________________________
const git_baseURL *{.strdefine.}= "https://github.com"
const git_owner   *{.strdefine.}= "UndefinedOwner"
const git_repo    *{.strdefine.}= "UndefinedRepository"


#_______________________________________
# @section Git Info: Types
#_____________________________
type Info * = object
  ## @descr Describes metadata/information about a Git repository, and provides tools to manage it.
  baseURL  *:string=  git_baseURL
  owner    *:string=  git_owner
  repo     *:string=  git_repo


#_______________________________________
# @section Git Info: Helpers
#_____________________________
const Templ_GitInfo * = "{info.baseURL}/{info.owner}/{info.repo}"
func `$` *(info :git.Info; templ :static string= Templ_GitInfo) :string= fmt( templ )
  ## @descr Converts the given git {@arg info} object into its corresponding URL

